File: .claude/herald/handlers/pre_tool_use.py
#!/usr/bin/env python3
"""PreToolUse hook - simplified security evaluation with Block-at-Commit strategy.

Block-at-Commit Strategy (from community best practices):
    "Intentionally do not use block-at-write hooks. Blocking an agent mid-plan
    confuses or even frustrates it."

    - ✅ Block at `git commit` if tests haven't passed
    - ❌ Don't block at Edit/Write operations
"""
from __future__ import annotations

import argparse
import json
import logging
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

from herald.utils.decision_api import DecisionAPI, DecisionResult
from herald.utils.constants import PRE_TOOL_USE, MAX_COMMAND_PREVIEW
from herald.utils.handler_result import HandlerResult

# Module-level logger
_log = logging.getLogger("hooks.pre_tool_use")
if not _log.handlers:
    _handler = logging.StreamHandler(sys.stderr)
    _handler.setFormatter(logging.Formatter("[%(name)s] %(levelname)s: %(message)s"))
    _log.addHandler(_handler)
    _log.setLevel(logging.WARNING)

# Commit gate file path (created by test runner when tests pass)
COMMIT_GATE_FILE = Path(os.getenv("HOOKS_COMMIT_GATE_FILE", "/tmp/hooks-herald-tests-pass"))


def _check_commit_gate(tool: str, command: str) -> tuple[bool, str | None]:
    """Check if git commit is allowed based on test gate.

    Block-at-Commit strategy: Only check at commit time, not during Edit/Write.
    Returns (is_allowed, reason_if_blocked).
    """
    # Only apply to Bash tool with git commit command
    if tool != "Bash":
        return True, None

    if "git commit" not in command:
        return True, None

    # Check if commit gate file exists (created when tests pass)
    if COMMIT_GATE_FILE.exists():
        return True, None

    # Allow if env var explicitly bypasses
    if os.getenv("HOOKS_SKIP_COMMIT_GATE", "").lower() in ("1", "true", "yes"):
        return True, None

    return False, "git commit blocked: run tests first (or set HOOKS_SKIP_COMMIT_GATE=1)"


def _preview_command(payload: dict[str, object]) -> str | None:
    """Extract command preview from payload."""
    cmd = payload.get("command")
    if isinstance(cmd, str) and cmd.strip():
        return cmd.strip()[:MAX_COMMAND_PREVIEW]
    args = payload.get("args")
    if isinstance(args, list) and args:
        joined = " ".join(str(a) for a in args)
        return joined[:MAX_COMMAND_PREVIEW] if joined else None
    return None


def _utc_timestamp() -> str:
    """Get current UTC timestamp in ISO format."""
    return datetime.now(timezone.utc).isoformat()


def handle_pre_tool_use(context) -> HandlerResult:
    """Handle pre tool use event - security evaluation before execution."""
    payload: dict[str, object] = context.payload if isinstance(context.payload, dict) else {}
    api: DecisionAPI = context.decision_api or DecisionAPI()

    issues: list[str] = []

    # Extract tool name
    tool_raw = (
        payload.get("tool")
        or payload.get("toolName")
        or payload.get("tool_name")
        or payload.get("name")
        or "unknown"
    )
    tool = str(tool_raw) if tool_raw else "unknown"
    if not tool or not tool.strip():
        tool = "unknown"
        issues.append("missing_tool_name")

    # Extract and parse tool input
    raw = payload.get("toolInput") or payload.get("tool_input") or payload.get("input")
    tool_input: dict[str, object] | None = None
    preview: str | None = None

    if isinstance(raw, dict):
        tool_input, preview = raw, _preview_command(raw)
    elif isinstance(raw, str):
        try:
            parsed = json.loads(raw)
            if isinstance(parsed, dict):
                tool_input, preview = parsed, _preview_command(parsed)
            elif isinstance(parsed, list):
                tool_input = {"args": parsed}
                preview = _preview_command(tool_input)
            else:
                issues.append("unsupported_tool_input_type")
                preview = raw[:MAX_COMMAND_PREVIEW]
        except json.JSONDecodeError:
            issues.append("invalid_tool_input_json")
            preview = raw[:MAX_COMMAND_PREVIEW]
    elif isinstance(raw, (list, tuple)):
        tool_input = {"args": [str(x) for x in raw]}
        preview = _preview_command(tool_input)
    elif raw is not None:
        issues.append("unsupported_tool_input_type")
        preview = str(raw)[:MAX_COMMAND_PREVIEW]

    # Make security decision
    try:
        # Provide empty dict if tool_input is None
        decision = api.pre_tool_use_decision(tool, tool_input or {})
    except Exception as exc:
        # Fallback to ask decision on API failure
        # Note: issues list will be added to additionalContext later
        decision = DecisionResult(
            "ask",
            "無法評估工具安全性，請人工確認",
            blocked=True,
        )
        # Store error info for later
        issues.append(f"api_error:{type(exc).__name__}")

    # If there are parsing issues, escalate to ask
    if not decision.blocked and issues:
        decision = DecisionResult(
            "ask",
            "工具輸入格式不明確，請人工確認",
            blocked=True,
        )

    # Block-at-Commit strategy: check if git commit is allowed
    if not decision.blocked and preview:
        commit_allowed, commit_reason = _check_commit_gate(tool, preview)
        if not commit_allowed:
            decision = DecisionResult(
                "ask",
                commit_reason or "Commit gate check failed",
                blocked=True,
            )
            issues.append("commit_gate_blocked")

    # Build additional context - always start fresh and merge
    existing_context = decision.payload.get("additionalContext")
    additional_context: dict[str, object] = {}

    # Merge existing context if present
    if isinstance(existing_context, dict) and existing_context:
        additional_context.update(existing_context)

    # Always set tool and issues
    additional_context["tool"] = tool
    if issues:
        additional_context["issues"] = issues

    # Build audit record
    audit: dict[str, object] = {
        "decision": decision.payload.get("permissionDecision") or decision.payload.get("decision", "unknown"),
        "blocked": decision.blocked,
        "timestamp": _utc_timestamp(),
    }
    if preview:
        audit["commandPreview"] = preview

    additional_context["preToolUseAudit"] = audit

    # Update decision payload
    decision.payload["additionalContext"] = additional_context

    # Build handler result
    hr = HandlerResult()
    hr.decision_payload = decision.to_dict()
    hr.continue_value = not decision.blocked

    perm = decision.payload.get("permissionDecision")
    if perm == "allow" and not decision.blocked:
        hr.suppress_audio = True
    else:
        hr.audio_type = PRE_TOOL_USE

    return hr


def main() -> int:
    """Entry point for manual invocations."""
    parser = argparse.ArgumentParser(description="Claude Code PreToolUse hook")
    parser.add_argument("--enable-audio", action="store_true",
                       help="Enable audio feedback")
    _ = parser.parse_args()

    try:
        payload = json.loads(sys.stdin.read().strip() or "{}")
    except Exception:
        payload = {}

    from herald.dispatcher import dispatch
    response = dispatch(PRE_TOOL_USE, payload)
    print(json.dumps(response))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: .claude/hooks/handlers/pre-tool-use.ts
/**
 * PreToolUse handler — security gating.
 * Audio only for deny/ask; clean allows are silent.
 */

import type { HookContext, HandlerResult, DecisionApi, DecisionResult } from "../lib/types";
import { createResult } from "../lib/types";
import { PRE_TOOL_USE } from "../lib/constants";
import { createDecisionApi } from "../lib/decision";

const MAX_COMMAND_PREVIEW = 240;

function previewCommand(payload: Record<string, unknown>): string | null {
  const cmd = payload.command;
  if (typeof cmd === "string" && cmd.trim()) {
    return cmd.trim().slice(0, MAX_COMMAND_PREVIEW);
  }
  const args = payload.args;
  if (Array.isArray(args) && args.length > 0) {
    const joined = args.map(String).join(" ");
    return joined ? joined.slice(0, MAX_COMMAND_PREVIEW) : null;
  }
  return null;
}

function utcTimestamp(): string {
  return new Date().toISOString();
}

export function handlePreToolUse(context: HookContext): HandlerResult {
  const payload = typeof context.payload === "object" && context.payload !== null ? context.payload : {};
  const api: DecisionApi = context.decisionApi ?? createDecisionApi();

  const issues: string[] = [];

  // Extract tool name
  let tool = String(
    payload.tool ?? payload.toolName ?? payload.tool_name ?? payload.name ?? "unknown",
  );
  if (!tool.trim()) {
    tool = "unknown";
    issues.push("missing_tool_name");
  }

  // Extract and parse tool input
  const raw = payload.toolInput ?? payload.tool_input ?? payload.input;
  let toolInput: Record<string, unknown> | null = null;
  let preview: string | null = null;

  if (raw !== null && raw !== undefined) {
    if (typeof raw === "object" && !Array.isArray(raw)) {
      toolInput = raw as Record<string, unknown>;
      preview = previewCommand(toolInput);
    } else if (typeof raw === "string") {
      try {
        const parsed = JSON.parse(raw);
        if (typeof parsed === "object" && parsed !== null && !Array.isArray(parsed)) {
          toolInput = parsed;
          preview = previewCommand(parsed);
        } else if (Array.isArray(parsed)) {
          toolInput = { args: parsed };
          preview = previewCommand(toolInput);
        } else {
          issues.push("unsupported_tool_input_type");
          preview = raw.slice(0, MAX_COMMAND_PREVIEW);
        }
      } catch {
        issues.push("invalid_tool_input_json");
        preview = raw.slice(0, MAX_COMMAND_PREVIEW);
      }
    } else if (Array.isArray(raw)) {
      toolInput = { args: raw.map(String) };
      preview = previewCommand(toolInput);
    } else {
      issues.push("unsupported_tool_input_type");
      preview = String(raw).slice(0, MAX_COMMAND_PREVIEW);
    }
  }

  // Evaluate safety
  let decision: DecisionResult;
  try {
    decision = api.preToolUseDecision(tool, toolInput ?? {});
  } catch (e) {
    decision = api.ask("無法評估工具安全性，請人工確認", PRE_TOOL_USE);
  }

  // If not blocked but has issues, escalate to ask
  if (!decision.blocked && issues.length > 0) {
    decision = api.ask("工具輸入格式不明確，請人工確認", PRE_TOOL_USE);
  }

  // Enrich payload
  const extra = (decision.payload.additionalContext ?? {}) as Record<string, unknown>;
  if (!extra.tool) extra.tool = tool;
  if (issues.length > 0 && !extra.issues) extra.issues = issues;
  decision.payload.additionalContext = extra;

  const audit: Record<string, unknown> = {
    decision: decision.payload.permissionDecision ?? decision.payload.decision ?? "unknown",
    blocked: decision.blocked,
    timestamp: utcTimestamp(),
  };
  if (preview) audit.commandPreview = preview;
  if (!decision.payload.preToolUseAudit) decision.payload.preToolUseAudit = audit;

  // Build result
  const hr = createResult();
  hr.decisionPayload = decision.toDict();
  hr.continueValue = !decision.blocked;

  const perm = decision.payload.permissionDecision;
  if (perm === "allow" && !decision.blocked) {
    hr.suppressAudio = true;
  } else {
    hr.audioType = PRE_TOOL_USE;
  }

  return hr;
}
